File: ELSE/SS/day6/kursal.py
n, m = map(int, input().split())
edges = []
for i in range(m):
    edges.append(list(map(int, input().split())))
edges.sort(key=lambda x: x[2])
parent=[i for i in range(n+1)]
def find(x):
    if x != parent[x]:
        parent[x] = find(parent[x])
    return parent[x]
def union(x,y):
    x ,y = find(x),find(y)
    if x != y:
        parent[x] = y
ans=0
cnt=0
for i in range(m):
    if (find(edges[i][0])!=find(edges[i][1])):
        ans+=edges[i][2]
        cnt+=1
        union(edges[i][0],edges[i][1])
    if cnt==n-1:
        break
print(ans)

File: ELSE/C+++C/SS/day6/kursal.cpp
#include <bits/stdc++.h>
using namespace std;
#define MAX 1000001
struct node
{
    int a, b, d;
} edges[MAX];
bool cmp(node a, node b)
{
    return a.d < b.d;
}
int parent[MAX];
int find(int x)
{
    int r = x;
    int i = x, j;
    while (parent[r] != r)
        r = parent[r];
    while (i != r)
    {
        j = parent[i];
        parent[i] = r;
        i = j;
    }
    return r;
}
int main()
{
    int n, m;
    cin >> n >> m;
    int a, b, d;
    int ans = 0;
    for (int i = 0; i < m; i++)
        scanf("%d %d %d", &edges[i].a, &edges[i].b, &edges[i].d);
    for (int i = 1; i <= n; i++)
        parent[i] = i;
    sort(edges, edges + m, cmp);
    int cnt = 0;
    for (int i = 0; i < m; i++)
    {
        int fa = find(edges[i].a), fb = find(edges[i].b);
        if (fa != fb)
        {
            parent[fa] = fb;
            ans += edges[i].d;
            cnt++;
        }
        if (cnt == n - 1)
            break;
    }
    cout << ans;
}